File: 146-lru-cache/lru-cache.py
class Node():
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.left = Node(0,0)
        self.right = Node(0,0)
        self.left.next = self.right
        self.right.prev = self.left
        self.store = {}

    def remove(self, n):
        p, nn = n.prev, n.next
        p.next = nn
        nn.prev = p
        
    
    def add(self, n):
        p = self.right.prev
        p.next = n
        n.prev = p
        self.right.prev = n
        n.next = self.right

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.store:
            temp = self.store[key]
            self.remove(temp)
            self.add(temp)
            return temp.value
        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.store:
            self.remove(self.store[key])
            del self.store[key]


        self.store[key] = Node(key, value)
        self.add(self.store[key])
        
        if len(self.store)  > self.capacity:
            del self.store[self.left.next.key]
            self.remove(self.left.next)

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.cpp
#include<unordered_map>

class Node{
    public:
    Node* prev = nullptr;
    Node* next = nullptr;
    int key;
    int value;

    public:
    Node(int key1, int value1){
        key = key1;
        value = value1;
    }
};

class LRUCache {
    int capacity;
    Node* left = new Node(0, 0);
    Node* right = new Node(0, 0);

    unordered_map<int, Node*> cache;

public:
    LRUCache(int capacity1) {
        capacity = capacity1;
        left->next = right;
        right->prev = left;
    }
    
    void add(Node* temp){
        Node* tempLeft = right->prev;
        tempLeft->next = temp;
        temp->prev = tempLeft;
        temp->next = right;
        right->prev = temp;
    }

    void remove(Node* temp){
        Node* tempLeft = temp->prev;
        Node* tempRight = temp->next;
        tempLeft->next = tempRight;
        tempRight->prev = tempLeft;
    }

    int get(int key) {
        if(cache.contains(key)){
            remove(cache[key]);
            add(cache[key]);
            return cache[key]->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(cache.contains(key)){
            remove(cache[key]);
            cache.erase(key);
        }
        Node* temp = new Node(key, value);
        cache[key] = temp;
        add(cache[key]);

        if(cache.size() > capacity){
            
            cache.erase(left->next->key);
            remove(left->next);
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */